File: app/src/main/java/com/veyndan/redditclient/ui/widget/ToggleableRadioButton.java
package com.veyndan.redditclient.ui.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.RadioButton;
import android.widget.RadioGroup;

/**
 * A {@link RadioButton} which can be deselected.
 */
public class ToggleableRadioButton extends RadioButton {

    public ToggleableRadioButton(final Context context) {
        super(context);
    }

    public ToggleableRadioButton(final Context context, final AttributeSet attrs) {
        super(context, attrs);
    }

    public ToggleableRadioButton(final Context context, final AttributeSet attrs,
                                 final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public void toggle() {
        if (isChecked()) {
            if (getParent() instanceof RadioGroup) {
                ((RadioGroup) getParent()).clearCheck();
            }
        } else {
            setChecked(true);
        }
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/ui/widget/ToggleableRadioButton.kt
package com.veyndan.paper.reddit.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.widget.RadioButton
import android.widget.RadioGroup

/**
 * A [RadioButton] which can be deselected.
 */
class ToggleableRadioButton : RadioButton {

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet,
                defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun toggle() {
        if (isChecked) {
            if (parent is RadioGroup) {
                (parent as RadioGroup).clearCheck()
            }
        } else {
            isChecked = true
        }
    }
}
